[OpenXR] Instance is never finalized even after exiting active sessions

svillar opened the pull request:
#### a976451bcf955ed04886a19be616a56417a96d08
<pre>
[OpenXR] Instance is never finalized even after exiting active sessions
<a href="https://bugs.webkit.org/show_bug.cgi?id=301515">https://bugs.webkit.org/show_bug.cgi?id=301515</a>

Reviewed by Fujii Hironori.

The OpenXR instance is now cleaned up once the session and its
associated resources are disposed. We don&apos;t need to keep it around
since creating new instances is a cheap operation.

No new tests required as there are already API tests verifying
the creation/destruction of OpenXR sessions.

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::~OpenXRCoordinator):
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::cleanupInstanceAndAssociatedResources):
(WebKit::OpenXRCoordinator::cleanupAllResources):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: <a href="https://commits.webkit.org/302232@main">https://commits.webkit.org/302232@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3190a44ee1186a7cb64d17387f61146b53fb423

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128311 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/587 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135702 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79780 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/461 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97664 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65573 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131259 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/352 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114947 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78255 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/335 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78991 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108719 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33538 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138158 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/432 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/407 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106202 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/471 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111288 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106002 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27032 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/347 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29838 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52735 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/483 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63636 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/382 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/451 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/447 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -638,6 +637,27 @@ void OpenXRCoordinator::cleanupSessionAndAssociatedResources()
+        return;
+
+    xrDestroyInstance(m_instance);
+    m_instance = XR_NULL_HANDLE;
```
`OpenXRCoordinator::startSession()` creates a new RenderState and a new WorkQueue by calling `Box<RenderState>::create()` and `WorkQueue::create()`.
`OpenXRCoordinator::endSessionIfExists()` immediately changes the state from Active to Idle.
If a session is started immediately after the previous session is requested to end, there is a race condition.

1. `startSession()` is called. WorkQueue A is created
2. `endSessionIfExists()`
3. `startSession()` is called. WorkQueue B is created
4. WorkQueue A calls `cleanupAllResources()` on XR_SESSION_STATE_EXITING

svillar replied:
Hmm I think there is no race there because `endSessionIfExists()` uses dispatchSync to first request the session exit and then it runs the render loop until there are not more events. By that point the XR_SESSION_STATE_EXITING should have been received and thus the cleanup happens before `endSessionIfExists()` finishes. Does it make sense?

fujii replied:
Oh, good point. I didn't realized dispatchSync is used.

svillar committed:
[OpenXR] Instance is never finalized even after exiting active sessions
https://bugs.webkit.org/show_bug.cgi?id=301515

Reviewed by Fujii Hironori.

The OpenXR instance is now cleaned up once the session and its
associated resources are disposed. We don't need to keep it around
since creating new instances is a cheap operation.

No new tests required as there are already API tests verifying
the creation/destruction of OpenXR sessions.

* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::~OpenXRCoordinator):
(WebKit::OpenXRCoordinator::startSession):
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::cleanupInstanceAndAssociatedResources):
(WebKit::OpenXRCoordinator::cleanupAllResources):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: https://commits.webkit.org/302232@main
